### PR TITLE
Add support for multiple backends

### PR DESCRIPTION
**Pull request by robin-nitrokey**
This is a reduced version of https://github.com/trussed-dev/trussed/pull/41 as discussed today.  The main difference is that the service endpoints cannot be configured at runtime (using a syscall), but only during client creation.  This makes the interface much simpler.  I did not convert the backend ID to an integer as originally planned because there is only one remaining struct that uses the backend ID, `ServiceEndpoint`.

Based on https://github.com/trussed-dev/trussed/pull/67

### CODE REVIEW

**Comment by robin-nitrokey** [comment]
This backends implementation only changes the public interface for a single function, `Service::add_endpoint`, which is probably not used outside Trussed anyway.  If this is a concern, we can introduce a variant to maintain compatibility.

**Comment by robin-nitrokey** [comment]
Rebased onto 88de4c6b3ac8c94534aad98efbed02af8eaec70c.

**Review by nickray — CHANGES_REQUESTED** [review]
This can be reverted, https://github.com/solokeys/solo2/pull/152 is merged. Thanks for that!

**Review by nickray — CHANGES_REQUESTED** [review]
I explored the suggestions concretely here: https://github.com/robin-nitrokey/trussed/compare/backends...nickray:trussed:elaboration

If this seems OK to you, happy to push to your branch and merge this PR.

**Comment by robin-nitrokey** [comment]
Thanks for the review!  While working on https://github.com/trussed-dev/trussed/pull/70 I realized that the current `Backends::select` method is not sufficient:  We have to add an associated type for the context to `Backend` which makes the trait non-object-safe, so we cannot return a reference to it.  Instead, we directly have to call the `Backend::reply_to` method for a given backend:

```rust
pub trait Backends<P: Platform> {
    type Id: 'static;
    type ExtensionId: TryFrom<u8, Error = Error>;
    type Context: Default;

    fn core_request(
        &mut self,
        id: &Self::Id,
        ctx: &mut Context<Self::Context>,
        request: &Request,
        resources: &mut ServiceResources<P>,
    ) -> Result<Reply, Error> {
        let _ = (id, ctx, request, resources);
        Err(Error::RequestNotAvailable)
    }

    fn extension_request(
        &mut self,
        backend: &Self::Id,
        extension: &Self::ExtensionId,
        ctx: &mut Context<Self::Context>,
        request: &request::Extension,
        resources: &mut ServiceResources<P>,
    ) -> Result<reply::Extension, Error> {
        let _ = (backend, extension, ctx, request, resources);
        Err(Error::RequestNotAvailable)
    }
}
```

So I would rather keep the `Backends` name or maybe use something more generic like `Dispatch`.

**Comment by nickray** [comment]
Dispatch sounds good. I find Backends too close to Backend. From the sound of it, though this PR needs to simmer a bit until #70 is more final?

**Comment by robin-nitrokey** [comment]
Okay, I’ll change it to `Dispatch`.  #70 would already be ready to review if this is merged.  I split it up into two PRs to make it easier to review, but if you’d prefer doing both in one PR, we can also do that.

**Comment by robin-nitrokey** [comment]
I’ve updated the PR:
- Rebased onto 1f3cb9db10c21a57f871725b339677c936a61b47
- Renamed `Backends` to `Dispatch` per our discussion
- Added a commit that introduces backend-specific context (extracted from #70)

**Review by daringer — APPROVED** [review]

File: src/client.rs [hunk 1]
```diff
@@ -685,6 +687,70 @@ pub trait UiClient: PollClient {
...
+    pub fn new(id: impl Into<PathBuf>) -> Self {
+        Self {
+            id: id.into(),
+            backends: &[],
+        }
+    }
+}
+
+impl<B: 'static> ClientBuilder<B> {
+    pub fn with_backends<C: 'static>(self, backends: &'static [BackendId<C>]) -> ClientBuilder<C> {
```
Comment (sosthene-nitrokey) [inline src/client.rs:705]: I would remove the `with`, it is generally not used in the builder pattern.

**Comment by robin-nitrokey** [comment]
Updated:
- Rebased onto 47afe55313ef418c9448b690c209a3c4f1c8fd64.
- Renamed `ClientContext` to `CoreContext`.
- Renamed `with_backends` in `ClientBuilder` to `backends`.

File: src/backend.rs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        ctx: &mut Context<Self::Context>,
+        request: &Request,
+        resources: &mut ServiceResources<P>,
+    ) -> Result<Reply, Error> {
+        let _ = (backend, ctx, request, resources);
+        Err(Error::RequestNotAvailable)
+    }
+}
+
+impl<P: Platform> Dispatch<P> for () {
```
Comment (nickray) [inline src/backend.rs:45]: I would prefer having this implemented on a specific type `struct SoftwareOnly` instead of using the unit type (name to be bikeshed). Think this is more clear.

File: src/backend.rs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    type BackendId = ();
```
Comment (nickray) [inline src/backend.rs:46]: I would prefer using a void type `enum Empty {}` here instead of using the unit type (name to be bikeshed).
Cf. https://github.com/robin-nitrokey/trussed/compare/backends...nickray:trussed:elaboration even if the Dispatch method changed now. This enforces that you can't even generate a BackendId::Custom.

File: src/backend.rs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    type Context = ();
```
Comment (nickray) [inline src/backend.rs:47]: Not sure what a "more correct" default type (empty struct vs empty enum) and name would be here.

File: src/client.rs [hunk 5]
```diff
@@ -685,6 +687,69 @@ pub trait UiClient: PollClient {
...
+            .map_err(|_service_endpoint| ())?;
+        Ok(requester)
+    }
+
+    #[allow(clippy::result_unit_err)]
+    pub fn build<P: Platform, D: Dispatch<P, BackendId = B>, S: Syscall>(
+        self,
+        service: &mut Service<P, D>,
+        syscall: S,
+    ) -> Result<ClientImplementation<S>, ()> {
```
Comment (nickray) [inline src/client.rs:729]: In these Results with trivial error types, maybe we should use Option and mark with `#[must_use]`?
I'm finding the proliferation of `()` a bit hard to follow.
Or would this break `?` in certain places?

**Review by nickray — CHANGES_REQUESTED** [review]

File: src/client.rs (hunk 5 above)
Comment (robin-nitrokey) [reply src/client.rs:729]: I just used `()` for consistency with the existing methods, but I agree that `Error` is much better.  I’ll add a `ClientCountExceeded` variant and change all relevant methods in `ClientBuilder` and `Service` to return that error instead.

File: src/backend.rs (hunk 4 above)
Comment (robin-nitrokey) [reply src/backend.rs:50]: I think the empty tuple is fine here because it just means “no data”.

File: src/backend.rs (hunk 3 above)
Comment (robin-nitrokey) [reply src/backend.rs]: Done.

Comment (robin-nitrokey) [reply src/backend.rs]: Could probably be `!` once it’s stable.

**Comment by robin-nitrokey** [comment]
Updated:
- Added `Error::ClientCountExceeded` variant and used it in the relevant methods.
- Added `SoftwareOnly` struct as the baseline `Dispatch` implementation.
- Added `Empty` enum.

I’m not sure whether “Software” is the best term to describe the Trussed default backend.  I think it was chosen with backends like SE050 in mind, but we will probably also have software-only backends that just implement extensions.  Maybe “Core” would be better?

File: src/backend.rs (hunk 4 above)
Comment (nickray) [reply src/backend.rs:50]: So `NoData`? :)

File: src/client.rs (hunk 5 above)
Comment (nickray) [reply src/client.rs]: Yeah that works. Generally it'd be better to have tight error types (i.e. a client of just this method would know which error variants it can actually expect), but doing this in Rust is tedious as you'd need tons of conversion traits.

**Comment by nickray** [comment]
I suggested `SoftwareOnly` for consistency with `BackendId::Software` (so change both if you change one). `Core` would be consistent with your `CoreContext`. `Default` or `Baseline` or whatever alternatives... Agree that the configured default backend doesn't necessarily have to be software, nor a custom backend non-software.

File: src/backend.rs [hunk 6]
```diff
@@ -0,0 +1,43 @@
+use crate::{
+    api::{Reply, Request},
+    error::Error,
+    platform::Platform,
+    service::ServiceResources,
+    types::{ClientContext, Empty},
+};
+
+pub enum BackendId<C> {
```
Comment (nickray) [inline src/backend.rs:9]: Can we have `I` instead of `C` here please?

File: src/client.rs [hunk 7]
```diff
@@ -685,6 +687,65 @@ pub trait UiClient: PollClient {
     }
 }
 
+pub struct ClientBuilder<B: 'static = Empty> {
```
Comment (nickray) [inline src/client.rs:690]: `I` instead of `B`. It's a bit tricky to stay on top of Backend vs Id throught this code, I think this would help.

File: src/client.rs [hunk 8]
```diff
@@ -685,6 +687,65 @@ pub trait UiClient: PollClient {
...
+    pub fn new(id: impl Into<PathBuf>) -> Self {
+        Self {
+            id: id.into(),
+            backends: &[],
+        }
+    }
+}
+
+impl<B: 'static> ClientBuilder<B> {
+    pub fn backends<C: 'static>(self, backends: &'static [BackendId<C>]) -> ClientBuilder<C> {
```
Comment (nickray) [inline src/client.rs:705]: Perhaps `J` instead of `C`?

File: src/client.rs [hunk 9]
```diff
@@ -685,6 +687,65 @@ pub trait UiClient: PollClient {
...
+impl ClientBuilder {
+    pub fn new(id: impl Into<PathBuf>) -> Self {
+        Self {
+            id: id.into(),
+            backends: &[],
+        }
+    }
+}
+
+impl<B: 'static> ClientBuilder<B> {
```
Comment (nickray) [inline src/client.rs:704]: `I` -> `B`

File: src/pipe.rs [hunk 10]
```diff
@@ -71,11 +72,12 @@ cfg_if::cfg_if! {
 // https://xenomai.org/documentation/xenomai-2.4/html/api/group__native__queue.html
 // https://doc.micrium.com/display/osiiidoc/Using+Message+Queues
 
-pub struct ServiceEndpoint {
+pub struct ServiceEndpoint<B: 'static = ()> {
```
Comment (nickray) [inline src/pipe.rs]: `I`

**Review by nickray — CHANGES_REQUESTED** [review]

**Comment by robin-nitrokey** [comment]
Updated:
- Changed `B` to `I` for the backend ID (and `C` to `J`).
- Added `NoData` struct (though I’m not really convinced ;)).
- Added documentation.
- Changed “software” to “core” when referring to the default request implementation in Trussed.

**Comment by robin-nitrokey** [comment]
Rebased onto 5c9bc60eaecf7d38b3512b973ac74309980ae721 and added to changelog.

File: src/types.rs [hunk 11]
```diff
@@ -233,17 +241,36 @@ pub mod consent {
...
+        }
+    }
+}
+
+// The "CoreContext" struct is the closest equivalent to a PCB that Trussed
 // currently has. Trussed currently uses it to choose the client-specific
 // subtree in the filesystem (see docs in src/store.rs) and to maintain
 // the walker state of the directory traversal syscalls.
-pub struct ClientContext {
+pub struct CoreContext {
```
Comment (nickray) [inline src/types.rs:267]: Shouldn't this be non-exhaustive in light of other discussions?

**Comment by daringer** [comment]
:+1:

File: src/types.rs (hunk 11 above)
Comment (robin-nitrokey) [reply src/types.rs:267]: I’ll add it to #71.  It has to be part of the public API so that the backends have access to it.
